### PR TITLE
Dial probes with proxy

### PR DESCRIPTION
**Pull request by cfryanr**
Fixes https://github.com/vmware-tanzu/pinniped/issues/2051.

- GitHubIdentityProvider will now avoid using `tls.Dial` to probe the connection for calculating status conditions by instead making a real HTTPS request to the GitHub API as its connection probe, because this request will respect the pod's proxy settings.
- WebhookAuthenticator will now detect the proxy setting and skip dialing the connection probe if the request should go through a proxy. The status conditions will inform the user that the check was skipped. For a webhook, we don't know anything about the expected behavior of the server except that we should be able to POST to the provided URL for authentication, but we don't want to post some fake data to the real authentication webhook endpoint just for the purposes of testing the connection. Until we have a better idea, we can just skip the probe when the provided URL would be impacted by the pod's proxy settings.
- Changed the URL configured for the test webhook to use the hostname `local-user-authenticator.local-user-authenticator.svc.cluster.local` instead of `local-user-authenticator.local-user-authenticator.svc` (added `.cluster.local`) because for some reason Squid cannot resolve domain names that end with `.svc`, even though `curl` works when shelled into Squid's pod container. Squid has several configuration options related to DNS resolution, but I couldn't figure out if any of those are the problem. Also added this DNS name to the SAN list when local-user-authenticator generates its TLS serving certificate, so it can be validated with either hostname.
- Added a new feature to `hack/prepare-for-integration-tests.sh` allowing the user to optionally firewall some of the IDPs and configure `HTTPS_PROXY`/`NO_PROXY` for the Supervisor and Concierge by simply setting the env variable `FIREWALL_IDPS=yes` when invoking this script. This allows running the integration tests to ensure that everything still works for `GitHubIdentityProvider`, `OIDCIdentityProvider`, `WebhookAuthenticator`, and `JWTAuthenticator` when the Supervisor or Concierge pods are accessing those servers through a web proxy. Note that setting `HTTPS_PROXY` does not apply to `LDAPIdentityProvider` or `ActiveDirectoryIdentityProvider` because the LDAP protocol does not work through HTTP proxies, since it is not based on HTTP. This works as expected on my laptop and both demonstrates the bug before the fix and demonstrates that the fix works.
- There is also a new CI job which firewalls the IDPs and enables the web proxy called `integration-test-idps-firewalled`. See CI repo for those changes. To make the NetworkPolicies function correctly in that CI job, I also had to upgrade the OS for the GCP VMs that run the kind clusters in CI to use the latest stable version of COS (`Google, Container-Optimized OS, 117-18613.0.76 stable, Kernel: COS-6.6.44 Kubernetes: 1.30.3 Docker: 24.0.9 Family: cos-stable`).
  - [Here](https://ci.pinniped.dev/teams/main/pipelines/main/jobs/integration-test-idps-firewalled/builds/5#L670857c0:50) is an example of it failing and demonstrating the bug for JWTAuthenticator on the main branch before this fix was merged. You can see that the JWTAuthenticator's status includes the message `cannot dial server: dial tcp 10.96.104.217:443: i/o timeout` because the NetworkPolicy prevents the Concierge pod from directly accessing the webhook server, and the code on main does not respect the HTTPS_PROXY setting while calculating status for the WebhookAuthenticator.
  - [Here](https://ci.pinniped.dev/teams/main/pipelines/main/jobs/integration-test-idps-firewalled/builds/5#L670857c0:1869) is a similar example of the failure on the main branch demonstrating the bug for GitHubIdentityProvider before this fix was merged. The status includes the message `cannot dial "api.github.com:443" for spec.githubAPI.host ("github.com"): dial tcp 140.82.112.5:443: i/o timeout` because the NetworkPolicy prevents the Supervisor pod from directly reaching GitHub and  the code on main does not respect the HTTPS_PROXY setting while calculating status for the GitHubIdentityProvider.
  - The same job passes for this PR, showing that this PR fixes and bug and that there are no other similar bugs in related parts of code. See link to this job in this PR's status checks shown below.

**Release note**:

```release-note
TBD
```

### CODE REVIEW

File: internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go [hunk 1]
```diff
@@ -493,18 +509,43 @@ func (c *gitHubWatcherController) validateGitHubConnection(
...
+	}, httpClient, nil
+}
+
+// ProbeURL is the production code for how to probe a GitHub URL to test our connection to the GitHub API.
+// It can be replaced via constructor injection for testing.
+func ProbeURL(ctx context.Context, client *http.Client, url string) error {
+	probeRequestCtx, probeRequestCancel := context.WithTimeout(ctx, 30*time.Second)
+	defer probeRequestCancel()
+
+	probeRequest, err := http.NewRequestWithContext(probeRequestCtx, http.MethodGet, url, nil)
```
Comment (cfryanr) [inline internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go:523]: Should we set the `X-GitHub-Api-Version`, `Accept`, and `User-Agent` headers on this request? See https://docs.github.com/en/rest/using-the-rest-api/getting-started-with-the-rest-api?apiVersion=2022-11-28#accept

Comment (cfryanr) [reply internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go:523]: The API docs for the rate limit API https://docs.github.com/en/rest/rate-limit/rate-limit?apiVersion=2022-11-28 suggest to use `Accept: application/vnd.github+json`:
```
curl -L \
  -H "Accept: application/vnd.github+json" \
  -H "Authorization: Bearer <YOUR-TOKEN>" \
  -H "X-GitHub-Api-Version: 2022-11-28" \
  https://api.github.com/rate_limit
```

Comment (cfryanr) [reply internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go:523]: Perhaps use these headers: `curl -L -H "Accept: application/vnd.github+json" -H "X-GitHub-Api-Version: 2022-11-28" -H "User-Agent: pinniped-concierge/v1" https://api.github.com/rate_limit`

Comment (cfryanr) [reply internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go:523]: Added in latest commit.

**Comment by cfryanr** [comment]
CI is passing, except for an unrelated problem with our AD server.

**Review by ashish-amarnath — APPROVED** [review]

**Pull request merged**